# Caesar ciphers
In this workshop, you'll write simple Python programs to encipher and decipher messages using the Caesar cipher. You'll also write a simple program to automatically break messages encrypted with a Caesar cipher.

## Using Jupyter Lab
This notebook is a series of cells. Some cells, like this one, contain text and pictures. Some cells contain code: they have a pair of square brackets to the left, like the `import string` cell just below. Some code cells are blank, for you to put your own bits of code in.

To _run_ a code cell, make sure the cell is selected (shown by a blue bar on the left) and press the 'play' triangle button on the toolbar above, or press Shift+Return. Jupyter will then run the code in that cell.

## Python pitfalls
Python can be very particular about some things. If some code you write doesn't work, check these things first.

* **Upper- and lower-case**. Python is case-sensitive, so the variables `letter`, `Letter`, and `LETTER` are all different. 
* **Brackets**. Make sure every open bracket has a matching close bracket. Jupyter's pretty good at inserting the close bracket automatically, so keep your eyes up while typing. And round brackets `()`, square brackets `[]`, and curly brackets `{}`, are all different, so don't mix them up.
* **Colons**. Some statements, like `for` and `if` have lines that end with colons. It's really easy to miss them out.
* **Layout**. Python really cares about layout. It shows control structure by intending code. Make sure line starts are vertically aligned.

These last two mean that your code should look something like this:

```python
def my_function():
    if some_condition:
        thing_inside_if()
        do_something()
    else:
        something_else()
    result = outside_the_if()
    another_thing()
    return result

my_answer = my_function()
my_answer        
```

Note the colons at the ends of the lines, the matching brackets, and how everything lines up in columns.

We'll be using the [`string` module](https://docs.python.org/3/library/string.html) from Python's standard library. That module defines some useful constants, such as `string.ascii_letters`, `string.ascii_lowercase` and `string.ascii_uppercase`.

Let's import it, so it's available for use.

In [1]:
import string

## What is a [Caesar cipher](https://en.wikipedia.org/wiki/Caesar_cipher)?

![360px-Cipher_device](360px-Cipher_device.jpg)

The [Caesar cipher](https://en.wikipedia.org/wiki/Caesar_cipher) is a simple cipher where each plaintext letter is shifted some positions along the alphabet to give the ciphertext letter. It can be used with two concentric wheel, like in the photo. You align the inner wheel with the appropriate part of the outer wheel. You can then find the plaintext letter in the inner wheel and the ciphertext letter is the corresponding one in the outer wheel.

Encryption happens letter by letter. Non-letter characters generally stay the same. 

Decryption works the other way, just in reverse. To find the plaintext letter, reverse the original shift from the ciphertext letter. 

For example, a shift of four would make the following conversions:

| Plaintext | A | B | C | D | E | F | G | H | I | J | K | L | M | N | O | P | Q | R | S | T | U | V | W | X | Y | Z |
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
| Ciphertext | E | F | G | H | I | J | K | L | M | N | O | P | Q | R | S | T | U | V | W | X | Y | Z | A | B | C | D |

The word `test` would be enciphered as `xiwx`, and `lippk` would be deciphered to `hello`.

## Implementing the Caesar cipher

How should we go about writing a program to do this? 

As each letter is enciphered independently, we can do a very simple function to encipher a message:

```
to ceasar-encipher (message, shift):
    for each character in message:
        caesar-encipher-letter (character, shift)
```

Similarly, we can exploit that the deciphering is just enciphering with a negative shift, and write a very simple function to decipher a message:

```
to caesar-decipher (message, shift):
    caesar-encipher (message, -shift)
```

How do we encipher a character? We need to be a bit *defensive* here, and account for the fact that we may be passed non-letters. We also have to think about how we want to handle upper- and lower-case letters. Should non-letters be removed from the result, kept unchanged, cause an error, or something else? Should all letters be converted to upper-case, or all converted to lowercase? 

I'll be conservative here, and say that non-letters are retained unchanged, and the case of letters should be retained. 

This means I can write `caesar-encipher-letter` as:

```
to caesar-encipher-letter(character, shift):
    if character is a letter:
        cipherletter = letter + shift
        if character is uppercase:
            return uppercase(cipherletter)
        else:
            return cipherletter
    else:
        return character
```

The only complication is the `letter + shift` bit: most programming languages will reject this statement as characters and numbers can't be added together.

The `ord()` function converts a character into a number, based on the ASCII code. The `chr()` function converts a number back into a character. Let's try them out.

In [2]:
ord('a')

97

In [3]:
ord('d')

100

In [4]:
ord('z')

122

In [5]:
chr(97)

'a'

In [6]:
chr(120)

'x'

But having numbers in the range 97–122 isn't very conventient. We can move the numbers to a more sensible range of 0–25 if we subtract the code of 'a'. like this:

In [7]:
ord('a') - ord('a')

0

In [8]:
# Teacher: Johnny, give me a sentence starting with 'I'.
# Johhny: "I is..."
# Teacher: No Johnny. Sentences start with "I" should be "I am".
# Johnny: OK. "I am the ninth letter of the alphabet".
#
# But as we're counting from zero, 'i' should convert to 8.
ord('i') - ord('a')

8

In [9]:
ord('z') - ord('a')

25

In [10]:
# If 'z' is 25, 24 should be 'y'
chr(24 + ord('a'))

'y'

As we'll be doing a lot of this sort of thing in future, let's define a couple of utility functions which convert letters to their positions in the alphabet, and back again. Because we'll be using these functions a lot, I give them deliberately short names. `pos` finds the position of a letter in the alphabet, ignoring case, and `raise`ing an error if it's given something other than a letter. 

Let's start with `pos()`, which converts a letter to a number.

In [11]:
def pos(letter):
    return ord(letter) - ord('a')

And let's try it.

In [12]:
pos('a')

0

In [13]:
pos('m')

12

In [14]:
pos('z')

25

Try a few of your own in the cells below.

That all seems to work. What about upper case letters?

In [15]:
pos('B')

-31

That's wrong. Let's fix it.

If we do the calculation differently, we get the right answer:

In [16]:
ord('B') - ord('A')

1

But how do we detect whether to subtract `ord('a')` or `ord('A')`?

Here's how to detect if a letter is upper- or lower-case.

In [17]:
def show_case(letter):
    if letter in string.ascii_lowercase:
        return 'lowercase'
    elif letter in string.ascii_uppercase:
        return 'uppercase'
    else:
        return 'not a letter'

In [18]:
show_case('a')

'lowercase'

In [19]:
show_case('A')

'uppercase'

In [20]:
show_case('!')

'not a letter'

## Your turn: writing `pos()`
Use the logic in `show_case()` to rewrite `pos()`. Your rewrite should detect if the letter is upper- or lower-case and return the right number. If it's not a letter, use the statement 
```python
raise ValueError('pos requires input of {} to be an ascii letter'.format(letter))
```

If you're stuck, you can [see the solution](pos-solution.ipynb).

In [21]:
# your pos here

## Testing `pos()`

Does your function work? Try some examples. Here are a few, but try some more:

(Note that Python uses `==` to test if things are equal; `=` is used for variable assignment.)

In [23]:
pos('a') == 0

True

In [24]:
pos('z') == 25

True

In [25]:
pos('m') == 12

True

## Converting back

`unpos()` does the reverse, convering a number into a (lowercase) letter. 

But sometimes, with a Caesar cipher, the shifted "letter" will be outside the range of the alphabet. For instance, if we encipher the word "yes" with a shift of 9, we get this:

| Letter | Number | Shifted number |
|:------:|:------:|:--------------:|
| y      | 24     | 33             | 
| e      |  4     | 13             |
| s      | 18     | 27             |

But 33 and 27 aren't in the range we want (0–25, remember). What we _want_ is this:

| Letter | Number | Shifted number | Converted number | Cipher letter |
|:------:|:------:|:--------------:|:----------------:|:-------------:|
| y      | 24     | 33             | 7                | h             |
| e      |  4     | 13             | 13               | n             |
| s      | 18     | 27             | 1                | b             |

The modulus operator, `%`, does that conversion. It find the remainder after a division. For instance, trims the number into that range. 

For instance, taking a number modulus 10 gives us just the "units" part of the number:

In [26]:
154 % 10

4

For our purposes, taking the number modulus 26 does the conversion for us. Try it out by seeing if the "shifted number"s above, taken modulus 26, give the "converted number" we want.

In [27]:
33 % 26

7

Numbers inside the range we want are unaffected by the modulus operator:

In [28]:
13 % 26

13

What about negative shifts? What about deciphering a ciphertext 'c' with a shift of 17?

In [29]:
pos('c') - 17

-15

In [30]:
(pos('c') - 17) % 26

11

## Your turn: writing `unpos()`
Using the examples above, write `unpos()`. It should take a number and convert it to the corresponding letter. If the number is outside the range 0–25, it should be converted to that range. 

* Hint: `chr()` converts a number to the character of that ASCII code.
* Hint: use `ord('a')` to convert the number to the ASCII code we want.
* Hint: use `ord('a')` _after_ you've done the modulus operation.

If you're stuck, you can [see the solution](unpos-solution.ipynb).

In [47]:
# Your solution here

## Testing `unpos()`

Try a few examples to see if it works.

In [32]:
unpos(0) == 'a'

True

In [33]:
unpos(25) == 'z'

True

In [34]:
unpos(-1) == 'z'

True

# Enciphering a letter
We can use these to write the `caesar_encipher_letter()` function. Remember the pseudocode we wrote:

```
to caesar-encipher-letter(character, shift):
    if character is a letter:
        cipherletter = letter + shift
        if character is uppercase:
            return uppercase(cipherletter)
        else:
            return cipherletter
    else:
        return character
```

# Your turn: writing `caesar_encipher_letter()`

The Python code should be almost the same as this. Note that your function should have underscores in its name, not the hyphens I used for pseudocode.

Follow the Python syntax for `if` and defining functions. Remember to end `if` and `def` lines with a colon, and to have the different lines align vertically, depending on what block they're part of.

If you're stuck, you can [see the solution](caesar_encipher-solution.ipynb).

In [ ]:
def caesar_encipher_letter(letter, shift):
    # Your solution here

In [35]:
def caesar_encipher_letter(letter, shift):
    """Encipher a letter, given a shift amount"""
    if letter in string.ascii_letters:
        cipherletter = unpos(pos(letter) + shift)
        if letter in string.ascii_uppercase:
            return cipherletter.upper()
        else:
            return cipherletter
    else:
        return letter   

## Testing `caesar_encipher_letter()`
We've seen that, with a shift of 4, we have the following:

| Plaintext | Ciphertext |
|:---------:|:----------:|
| t | x |
| e | i |
| s | w | 
| h | l |
| l | p | 
| o | k |

Does your `caesar_encipher_letter()` do that?

In [36]:
caesar_encipher_letter('t', 4) == 'x'

True

# Enciphering messages
And finally, we can write `caesar_encipher` and `caesar_decipher`.

Note the syntax of `encipher += caesar_encipher_letter(character, shift)`. The `+=` is a bit of shorthand for 'increment', so 

```
encipher += caesar_encipher_letter(character, shift)
```
means the same as 
```
encipher = encipher + caesar_encipher_letter(character, shift)
```

You can do similar things with other operators, such as `*`, so

`item = item + change` and `item += change` are the same, 

and 

`item = item * change` and `item *= change` are the same.

In [37]:
def caesar_encipher(message, shift):
    """Encipher a message with the Caesar cipher of given shift"""
    enciphered = ""
    for character in message:
        enciphered += caesar_encipher_letter(character, shift)
    return enciphered

In [38]:
def caesar_decipher(message, shift):
    """Decipher a message with the Caesar cipher of given shift"""
    return caesar_encipher(message, -shift)

## Testing
Now let's see if it works:

In [39]:
caesar_encipher('This is a test message.', 4)

'Xlmw mw e xiwx qiwweki.'

In [40]:
caesar_decipher('Xlmw mw e xiwx qiwweki.', 4)

'This is a test message.'

Success!

Now encipher and decipher your own messages. Send them to your friends. Can they break the codes?

In [part 2](2-breaking-ciphers.ipynb) of this workshop, you'll see how you can get a computer to automatically break these ciphers for you.